HSEARCH-1809 Introduce the context for the massindexer batch thread work (#3397)
hibernate/hibernate-search: documentation/src/main/asciidoc/reference/indexing-massindexer.asciidoc

Implementations of `MassIndexingFailureHandler` must be thread-safe.

> |`massIndexingEnvironment(MassIndexingEnvironment)`

Reviewer: If we want to be consistent with other pluggable components (monitor, failure handler) we should not use the prefix for this method, just for the interface.

```suggestion
|`environment(MassIndexingEnvironment)`
```

Author: yeah, I had that intent initially, but `environment` seems to be too generic and sounds "too powerful", but if it's just me then I'm happy to make a change 😄

> |A mass indexing environment that has no additional actions configured.

Reviewer: ```suggestion
|An empty environment (no threadlocals, ...).
```

|

> The component responsible for setting before and after actions for the threads doing mass indexing.

Reviewer: ```suggestion
The component responsible for setting up an environment (threadlocals, ...) on mass indexing threads before mass indexing starts, and tearing down that environment after mass indexing.
```

> Implementations should handle their exceptions unless it is an unrecoverable situation in which further mass indexing
> does not make sense.

Reviewer: ```suggestion
Implementations should handle their exceptions unless it is an unrecoverable situation in which further mass indexing
does not make sense: any exception thrown by the `MassIndexingEnvironment` will abort mass indexing.
```
